### ISSUE TITLE
No way to override ignore patterns IN JS API

### ISSUE THREAD

**Issue by JounQin** [prettier/prettier#10395]
**Environments:**

- Prettier Version: 2.2.1
- Running Prettier via: 
- Runtime: 
- Operating System: 

**Steps to reproduce:**

**Expected behavior:**

**Actual behavior:**

**Comment by JounQin** [comment]
Sorry I'm editing on mobile, just accidentally post it, I'll fulfill it soon.

**Comment by thorn0** [comment]
`prettier.format()` doesn't check `.prettierignore` and overall knows nothing about ignoring files. Your problem must be somewhere else.

**Comment by JounQin** [comment]
Hah, you're right, I just forgot I've just used `prettier --write` for simple testing, but there is no warning about file been ignore.

I'd recommend adding a warning log like eslint:

```log
warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override
```

A `--no-ignore` flag for cli would be great.

**Comment by thorn0** [comment]
Git doesn't show warnings about files ignored because of `.gitignore`. Why should Prettier?
Please describe the use case. When and why exactly do you propose to show warnings?

> A --no-ignore flag for cli would be great.

`--ignore-path=/dev/null` works. Besides, what's the point in configuring `.prettierignore` first and then trying to disable it? Looks like too much of an edge case.

**Comment by JounQin** [comment]
> Git doesn't show warnings about files ignored because of .gitignore.

If you try to add some file which is ignored, or add a new ignore pattern and delete related files, git will warn.

> Please describe the use case. When and why exactly do you propose to show warnings?

Although some test files are ignored, but I want to use cli rather than node API to check if my changes work as expected quickly. A git diff change is much clean to review.

And also, sometimes I want to format first then change the test case content for better readability.

> --ignore-path=/dev/null works

This is not too intuitive, and does it work on Windows? I don't know.

**Comment by thorn0** [comment]
If you have `semi: false` in the config file, you don't expect Prettier to print warnings like "Semicolons aren't printed because...". Is `.prettierignore` different? You've configured a program. Now it works as configured. Why show warnings?

Sorry, I don't really understand your description of your use case and the scenario where you propose to show warnings. I hoped you would write exact steps.

E.g. `prettier -w src/myfile.js`. Should this, in your opinion, show a warning if `src/myfile.js` is in `.prettierignore`? What about `prettier --write src/*.js`? `prettier --write src`? `prettier --check src/myfile.js`?

As for "not intuitive", you're using the CLI in a non-intended way. I'm just suggesting how this can be done. But why should this be intuitive? It this something generally useful for a large enough number of Prettier users?

**Comment by JounQin** [comment]
ESLint prints a warning for every file separately.

A warning like ESLint's `warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override` is just enough, I don't think we need to read options.

My point is the file is ignored, but I run `prettier --write` expectedly or unexpectedly, a warning would always help.

**Comment by thorn0** [comment]
What about this part?

> E.g. prettier --write src/myfile.js. Should this, in your opinion, show a warning if src/myfile.js is in .prettierignore? What about prettier --write src/*.js? prettier --write src? prettier --check src/myfile.js?

Do you want to see a separate warning for every ignored file that matches the glob on running `--write src/*.js` or inside the `src` directory on running `prettier --write src`? What about `--check`?

**Comment by JounQin** [comment]
> What about this part?
> 
> > E.g. prettier --write src/myfile.js. Should this, in your opinion, show a warning if src/myfile.js is in .prettierignore? What about prettier --write src/*.js? prettier --write src? prettier --check src/myfile.js?
> 
> Do you want to see a separate warning for every ignored file that matches the glob on running `--write src/*.js` or inside the `src` directory on running `prettier --write src`? What about `--check`?

Didn't I answer this question? 😂

Personally I like the ESLint's way to warn about each file, the logic is just simple. And `prettier --check` reads `--ignore-path` too, right? So of course it should be affected also.

**Comment by thorn0** [comment]
ESLint doesn't warn about each file. I've just tested that. `--check` is meant for CI. Warnings on CI when everything works as expected doesn't sound like a good thing. That's actually what worries me here the most. Showing warnings for expected behavior.

**Comment by JounQin** [comment]
> ESLint doesn't warn about each file

![image](https://user-images.githubusercontent.com/8336744/108766175-c58c2000-758f-11eb-8481-908915b93ef4.png)

Weird?

**Comment by thorn0** [comment]
Potentially, I agree that adding a warning makes sense in two cases.

1. `prettier --write file.js` The file name was specified explicitly, but this file is ignored.
2. `prettier --write glob` if all the matching files are ignored.

That's what ESLint does if I'm not mistaken.

**Comment by JounQin** [comment]
> Potentially, I agree that adding a warning makes sense in two cases.
> 
> 1. `prettier --write file.js` The file name was specified explicitly, but this file is ignored.
> 2. `prettier --write glob` if all the matching files are ignored.
> 
> That's what ESLint does if I'm not mistaken.

Seems true to me.

**Comment by thorn0** [comment]
> Weird?

Yep. I tested `eslint <dirName>`, not `eslint <glob>`.

**Comment by JounQin** [comment]
> --check is meant for CI

If the user run `--check` on CI for ignored files, I think we should warn it because it seems to be wrong.

**Comment by JounQin** [comment]
> > Weird?
> 
> Yep. I tested `eslint <dirName>`, not `eslint <glob>`.

![image](https://user-images.githubusercontent.com/8336744/108766693-709cd980-7590-11eb-8410-0980310cca66.png)

**Comment by thorn0** [comment]
That's the "all the matching files are ignored" case. What about "some but not all"?

**Comment by thorn0** [comment]
> If the user run --check on CI for ignored files, I think we should warn it because it seems to be wrong.

No. Most of the time Prettier CLI is run this way: `prettier .` or `prettier src dir2 dir3`. We definitely don't want show warnings for ignored files in these cases.

**Comment by JounQin** [comment]
> What about "some but not all"?

Seems it just pass, and I think it's reasonable and acceptable, which is compatible with the common usage for `--check` you mentioned.

So `--write` and `--check` files are all ignored, or using glob, a warning should be emitted?

**Comment by thorn0** [comment]
Yes, I guess. There is also the question whether such a warning can break someone's setup. Should it be considered a breaking change?

**Comment by JounQin** [comment]
I personally won't say it is a breaking change, it's just a correct think to me.

Of course, we should listen some more voices.
